manager/integration/tests/test_scheduling.py

```diff
@@ -377,3 +377,50 @@ def test_replica_rebuild_per_volume_limit():
 
     """
     pass
+
+
+@pytest.mark.skip(reason="TODO")
+def test_data_locality_basic():
+    """
+    Test data locality basic feature
+
+    Context:
+
+    Data Locality feature allows users to have an option to keep a local
+    replica on the same node as the consuming pod.
+    Longhorn is currently supporting 2 modes:
+    - disabled: Longhorn does not try to keep a local replica
+    - best-effort: Longhorn try to keep a local replica
+
+    See manual test plan at
+    https://github.com/longhorn/longhorn/issues/1045#issuecomment-680706283
+
+    Steps:
+
+    1. Create a volume(1) with 1 replica and dataLocality set to disabled
+    2. Find node where the replica is located on.
+       Let's call the node is replica-node
+    3. Attach the volume to a node different than replica-node.
+       Let call the node is engine-node
+    4. Use a retry loop to verify that Longhorn does not create
+       a replica on the engine-node
+    5. Update dataLocality to best-effort for volume(1)
+    6. Use a retry loop to verify that Longhorn creates and rebuilds
+       a replica on the engine-node and remove the other replica
+    7. detach the volume(1) and attach it to a different node.
+       Let's call the new node is new-engine-node and the old
+       node is old-engine-node
+    8. Wait for volume(1) to finish attaching
+    9. Use a retry loop to verify that Longhorn creates and rebuilds
+       a replica on the new-engine-node and remove the replica on
+       old-engine-node
+    10. Update dataLocality to disabled for volume(1)
+    11. detach the volume(1) and attach it to a different node.
+       Let's call the new node is new-engine-node and the old
+       node is old-engine-node
+    12. Wait for volume(1) to finish detaching
+    13. Use a retry loop to verify that Longhorn does not create
+        a replica on the new-engine-node
+    14. clean up
```

Comment:
What about the case that replica cannot be created on the attached node?

Also, what if the volume detached (workload shutdown) during the time of detaching? Need to think more about negative cases.

Reply:
Thank you! I will fix this

Reply:
@yasker I added the 2 negative test cases and also check for replica deletion preference.